tty/imx6ull-uart/imx6ull-uart.c

```diff
@@ -145,48 +286,104 @@ void uart_thr(void *arg)
-	*(uart.base + ucr1) &= ~0x2000;
+	uint8_t flags = 0;
+	uint32_t ctrl1 = *(uart.base + ucr1);
+	uint32_t stat1 = *(uart.base + usr1);
+	uint32_t stat2 = *(uart.base + usr2);
+
+	/* check for errors */
+	if (stat1 & USR1_PARITYERR) {
+		*(uart.base + usr1) = USR1_PARITYERR;
+		uart.counters.parity++;
+		flags |= PARITY_ERR;
+	}
+	if (stat1 & USR1_FRAMERR) {
+		*(uart.base + usr1) = USR1_FRAMERR;
+		uart.counters.frame++;
+		flags |= FRAME_ERR;
+	}
+	if (stat2 & USR2_ORE) {
+		*(uart.base + usr2) = USR2_ORE;
+		uart.counters.hw_overrun++;
+		flags |= HW_OVERRUN_ERR;
+	}
+
+	/* RX */
+	if ((ctrl1 & UCR1_RRDYEN) && (stat1 & USR1_RRDY)) {
+		while ((*(uart.base + usr2) & USR2_RDR)) {
+			/* NOTE: lock-free push */
+			if (lf_fifo_push(uart.rx_sw_fifo, *(uart.base + urxd))) {
+				flags |= RX_DONE;
+			}
+			else {
+				uart.counters.sw_overrun++;
+				flags |= SW_OVERRUN_ERR;
+			}
+		}
+	}
+
+	/* TX */
+	if ((ctrl1 & UCR1_TRDYEN) && (stat1 & USR1_TRDY)) {
+		while (!fifo_is_empty(uart.tty_common.tx_fifo) && !(*(uart.base + uts) & UTS_TXFULL)) {
+			/* FIXME: potential data race on tx_fifo (lock-free access) */
+			*(uart.base + utxd) = fifo_pop_back(uart.tty_common.tx_fifo);
+			flags |= TX_DONE;
+		}
+
+		if (fifo_is_empty(uart.tty_common.tx_fifo))
+			*(uart.base + ucr1) &= ~UCR1_TRDYEN;
+	}
+
+	__atomic_or_fetch(&uart.int_flags, flags, __ATOMIC_RELAXED);
+
+	return (flags ? 0 : -1);
+}
+
+
+static void uart_process_rx(void)
+{
+	uint8_t c;
 
-	return uart.cond;
+	/* NOTE: lock-free pop */
+	while (lf_fifo_pop(uart.rx_sw_fifo, &c))
+		libtty_putchar(&uart.tty_common, c, NULL);
 }
 
 
 static void uart_intrthr(void *arg)
 {
-	for (;;) {
-		/* wait for character or transmit data */
-		mutexLock(uart.lock);
-		while (!(*(uart.base + usr2) & (1 << 0))) {  // nothing to RX
-			if (libtty_txready(&uart.tty_common)) { // we something to TX
-				if ((*(uart.base + usr1) & (1 << 13))) // TX ready
-					break;
-				else
-					*(uart.base + ucr1) |= 0x2000; // wait for TRDY interrupt
-			}
-			condWait(uart.cond, uart.lock, 0);
-		}
-		/* disable tx ready interrupt again (sticky conds) */
-		*(uart.base + ucr1) &= ~0x2000;
-
-		mutexUnlock(uart.lock);
+	uint8_t flags;
 
-		/* RX */
-		while ((*(uart.base + usr2) & (1 << 0)))
-			libtty_putchar(&uart.tty_common, *(uart.base + urxd), NULL);
+	mutexLock(uart.lock);
 
-		/* TX */
-		while (libtty_txready(&uart.tty_common)) {
-			if (*(uart.base + uts) & (1 << 4)) { // check TXFULL bit
-				break; /* wait in main loop for TX to be ready before resuming operation */
-			}
-			*(uart.base + utxd) = libtty_getchar(&uart.tty_common, NULL);
-		}
+	for (;;) {
+		do {
+			flags = __atomic_exchange_n(&uart.int_flags, 0, __ATOMIC_RELAXED);
+
+			if (flags & PARITY_ERR)
+				log_warn("parity error (%u)", uart.counters.parity);
+			if (flags & FRAME_ERR)
+				log_warn("frame error (%u)", uart.counters.frame);
+			if (flags & HW_OVERRUN_ERR)
+				log_warn("hw_overrun (%u)", uart.counters.hw_overrun);
```

Comment:
not sure if printing more chars (possibly on UART) on overflow is the best thing to do ;).

Side note: MISRA says we always need to use `{ }`